synapse/lib/auth.py

```diff
@@ -81,3 +97,900 @@ def runas(user):
+        robj = self.auth.reqRole(name)
+        robj.addRule(rule)
+        ret = s_tufo.tufo(mname, role=(name, robj._getAuthData()))
+        return True, ret
+
+    @reqAdmin
+    def __authDelRoleRule(self, mesg):
+        mname, mdict = mesg
+        name = mdict.get('role')
+        rule = mdict.get('rule')
+        robj = self.auth.reqRole(name)
+        robj.delRule(rule)
+        ret = s_tufo.tufo(mname, role=(name, robj._getAuthData()))
+        return True, ret
+
+    @reqAdmin
+    def __authAddAdmin(self, mesg):
+        mname, mdict = mesg
+        name = mdict.get('user')
+        uobj = self.auth.reqUser(name)
+        uobj.setAdmin(True)
+        ret = s_tufo.tufo(mname, user=(name, uobj._getAuthData()))
+        return True, ret
+
+    @reqAdmin
+    def __authDelAdmin(self, mesg):
+        mname, mdict = mesg
+        name = mdict.get('user')
+        uobj = self.auth.reqUser(name)
+        uobj.setAdmin(False)
+        ret = s_tufo.tufo(mname, user=(name, uobj._getAuthData()))
+        return True, ret
+
+    @reqAdmin
+    def __authAddUserRole(self, mesg):
+        mname, mdict = mesg
+        name = mdict.get('user')
+        role = mdict.get('role')
+        uobj = self.auth.reqUser(name)
+        robj = self.auth.reqRole(role)
+        uobj.addRole(role)
+        ret = s_tufo.tufo(mname, user=(name, uobj._getAuthData()))
+        return True, ret
+
+    @reqAdmin
+    def __authDelUserRole(self, mesg):
+        mname, mdict = mesg
+        name = mdict.get('user')
+        role = mdict.get('role')
+        uobj = self.auth.reqUser(name)
+        robj = self.auth.reqRole(role)
+        uobj.delRole(role)
+        ret = s_tufo.tufo(mname, user=(name, uobj._getAuthData()))
+        return True, ret
+
+class Auth(s_config.Config):
+    '''
+    An authorization object which can help enforce cortex rules.
+
+    Args:
+        dirn (str): Dictionry backing the Auth data.
+        conf (dict): Optional configuration data.
+    '''
+    def __init__(self, dirn, conf=None):
+
+        s_config.Config.__init__(self, opts=conf)
+
+        path = os.path.join(dirn, 'auth.lmdb')
+
+        mapsize = self.getConfOpt('lmdb:mapsize')
+
+        self.lenv = lmdb.open(path, max_dbs=128)
+        self.lenv.set_mapsize(mapsize)
+
+        self.onfini(self.lenv.close)
+
+        self._db_users = self.lenv.open_db(b'users')
+        self._db_roles = self.lenv.open_db(b'roles')
+
+        # these may be used by Auth() callers
+        self.users = {}
+        self.roles = {}
+
+        with self.lenv.begin() as xact:
+
+            for name, info in self._iterAuthDefs(xact, self._db_roles):
+                self.roles[name] = Role(self, name, info)
+
+            for name, info in self._iterAuthDefs(xact, self._db_users):
+                self.users[name] = User(self, name, info)
+
+    def initConfDefs(self):
+        self.addConfDefs((
+            ('lmdb:mapsize', {'type': 'int', 'defval': s_const.gigabyte,
```

Comment:
This should use the default valu from s_lmdb if we merge @jnwatson 's PR first.

Reply:
holding this till master merge